### PR TITLE
[SPMD] Upstream iter_move_grads_and_optimizers

### PR DESCRIPTION
**Pull request by fegin**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #98785

This PR upstreams `iter_move_grads_and_optimizer` which delay some of the gradients and the corresponding optimizer to the next iteration. D44512863(credit to @lessw2020 ) is the internal implementation, which is only good for the old _SPMD expansion.  This PR changes the implmentation to use the new APIs.

Differential Revision: [D44836486](https://our.internmc.facebook.com/intern/diff/D44836486/)

### CODE REVIEW

File: test/distributed/_spmd/test_transformation.py [hunk 1]
```diff
@@ -228,6 +231,37 @@ def train_step(model, optim, batch):
...
+            model(batch).sum().backward()
+            optim.step()
+            optim.zero_grad()
+
+        model, optim, _, _ = self._init(
+            batch_size, layers, dim, foreach=False, fused=True
+        )
+        for _ in range(num_iters):
+            batch = torch.randn(batch_size, dim).cuda()
+            out = train_step(model, optim, batch)
```
Comment (mrshenli) [inline test/distributed/_spmd/test_transformation.py:263]: same comment regarding accuracy tests as previous PRs

Comment (fegin) [reply test/distributed/_spmd/test_transformation.py]: Done

File: torch/distributed/_spmd/graph_optimization.py [hunk 2]
```diff
@@ -890,3 +891,49 @@ def split_fused_optimizer(
...
+    """Function to extract a comm block and split out a new optimizer and step for it.
+    This subgraph is then moved to the forward graph.
+    """
+    for comm_block in get_all_comm_blocks(gm, "all_reduce"):
+        if comm_block.comm_node.name == target_comm_node:
+            break
+    else:
+        raise ValueError(f"Cannot find {target_comm_node}")
+
+    optim_blocks = get_all_fused_optimizer_blocks(gm, "_fused_adam")
```
Comment (mrshenli) [inline torch/distributed/_spmd/graph_optimization.py:911]: this is very specific to adam. Let's add a TODO/FIXME to track

File: torch/distributed/_spmd/graph_utils.py [hunk 3]
```diff
@@ -104,6 +104,20 @@ def get_output(graph: fx.Graph) -> fx.Node:
...
+def find_node(
+    graph: fx.Graph, predicate: Callable, reverse_order: bool = False
+) -> List[fx.Node]:
+    """
+    Take a predicate and return all the nodes in the `graph` where the predicate
+    holds.
+    """
+    nodes = graph.nodes
+    if reverse_order:
+        nodes = reversed(nodes)
```
Comment (mrshenli) [inline torch/distributed/_spmd/graph_utils.py:116]: nit: simplification?

```python
nodes = reversed(graph.nodes) if reverse_order else graph.nodes
```

**Review by mrshenli — APPROVED** [review]
This is not user-facing yet, stamp to unblock. Please fix Lint before land

File: torch/distributed/_spmd/graph_optimization.py [hunk 4]
```diff
@@ -890,3 +891,49 @@ def split_fused_optimizer(
         return _split_fused_adam(gm, optim_block, split_gradients)
     else:
         raise NotImplementedError("Only fused_adam is supported now")
+
+
+@graph_optimization_pass(run_after=("schedule_comm_wait", "remove_copy_from_optimizer"))
+def iter_move_grads_and_optimizers(
+    gm: IterGraphModule,
+    target_comm_node: str,
+    target_dest_node: str,
```
[the remark this answers is not in the document]
Comment (fegin) [reply torch/distributed/_spmd/graph_optimization.py:900]: You are right about the usage of this pass. This enables the ability to move nodes around but the algorithm is not decided yet. Right now the API is intended to be used for manual tuning.

File: torch/distributed/_spmd/graph_utils.py (hunk 3 above)
Comment (fegin) [reply torch/distributed/_spmd/graph_utils.py:116]: This is less easier to read after typing hints. Will keep the original form.

**Pull request closed without merging**